### PR TITLE
website/blog: sync objects in external databases

### PR DESCRIPTION
**Pull request by a8m**
cc @masseelch, @giautm, @rotemtam, @yonidavidson, @hilakashai, @hedwigz

### CODE REVIEW

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+				return nil, errors.New("avatar field is missing")
+			}
+			// TODO:
+			// 1. Verify that "avatarURL" points to a real object in the bucket.
+			// 2. Otherwise, fail.
+			return next.Mutate(ctx, m)
+		})
+	}
+	// Limit the hook only to "Create" operations.
+	return hook.On(hk, ent.OpCreate)
```
Comment (yonidavidson) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:128]: Nice, in the examples usually the condition for the hook is on the schema (and not on the hook) but that's just semantics.

**Review by yonidavidson — APPROVED** [review]
I like it!

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+Ensuring consistency between two separate data systems is not a simple task.  When we want to propagate, for example,
+deletion of a record in one system to another, there is no obvious way to guarantee that the two systems will end in a
+synchronized state, since one of them may fail and the network link between them may be slow or down. Having said that,
+and especially with the prominence of microservices-architectures these problems have become more common, and
+distributed systems researchers have come up with patterns to solve them, such as the
+[Saga Pattern](https://microservices.io/patterns/data/saga.html).
+
+The application of these patterns is usually complex and difficult, and so in many cases architects do not go after a
+"perfect" and instead go after simpler solutions that involve either the acceptance of some inconsistency between the systems or background reconciliation procedures.
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:24]: linebreak

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+references to blobs in object storages such as AWS S3 or Google Cloud Storage.
+
+Ensuring consistency between two separate data systems is not a simple task.  When we want to propagate, for example,
+deletion of a record in one system to another, there is no obvious way to guarantee that the two systems will end in a
+synchronized state, since one of them may fail and the network link between them may be slow or down. Having said that,
+and especially with the prominence of microservices-architectures these problems have become more common, and
+distributed systems researchers have come up with patterns to solve them, such as the
+[Saga Pattern](https://microservices.io/patterns/data/saga.html).
+
+The application of these patterns is usually complex and difficult, and so in many cases architects do not go after a
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:23]: `a "perfect"` what? Or maybe `the "perfect"`

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+### Propagate Mutations to External Systems
+
+In our example, we are going to create a simple `User` schema with 2 immutable string fields, `"name"` and
+`"avatar_url"`. Let's run the `ent init` command for creating a skeleton schema for our `User`:
+
+```shell
+go run entgo.io/ent/cmd/ent init User
+```
+
+Then, add the `name` and the `avatar_url` fields and run `go generate` for generating the assets.
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:39]:
```suggestion
Then, add the `name` and the `avatar_url` fields and run `go generate` to generate the assets.
```

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+// Open a bucket named photos in Google Cloud Storage.
+if bucket, err := blob.OpenBucket(ctx, "gs://my-bucket"); err != nil {
+	log.Fatal("failed opening gs bucket:", err)
+}
+```
+
+### Schema Hooks
+
+[Hooks](https://entgo.io/docs/hooks) are a powerful feature of Ent that allows adding custom logic before and after
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:93]:
```suggestion
[Hooks](https://entgo.io/docs/hooks) are a powerful feature of Ent that allow adding custom logic before and after
```

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+operations that mutate the graph.
+
+Hooks can be either defined dynamically using `client.Use` (called "Runtime-hooks"), or explicitly on the schema
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:96]: `Runtime-hooks` and `Schema Hooks`. Either use dash or don't but don't mix.

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+}
+```
+
+Next, re-run code generation:
+
+```shell
+go generate ./ent
+```
+
+We can now can access the Bucket API from all generated builders. Let's finish the implementations of the above hooks.
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:181]:
```suggestion
We now can access the Bucket API from all generated builders. Let's finish the implementations of the above hooks.
```

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+	// Output:
+	// ["a8m.png"]
+	// []
+}
+```
+
+### Wrapping Up
+
+Great! We have configured Ent to extend our generated code and inject the `blob.Bucket` as an 
+[External Dependency](https://entgo.io/docs/code-gen#external-dependencies). Next, we defined 2 mutation hooks and used
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:293]:
```suggestion
[External Dependency](https://entgo.io/docs/code-gen#external-dependencies). Next, we defined two mutation hooks and used
```

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+the `blob.Bucket` API to ensure our product constraints are satisfied.
+
+The code for this example is available on [github.com/a8m/ent-sync-example](https://github.com/a8m/ent-sync-example).
```
Comment (masseelch) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:296]:
```suggestion
The code for this example is available at [github.com/a8m/ent-sync-example](https://github.com/a8m/ent-sync-example).
```

**Review by masseelch — APPROVED** [review]
Much like. Very good!

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+author: Ariel Mashraki
+authorURL: https://github.com/a8m
+authorImageURL: "https://avatars0.githubusercontent.com/u/7413593"
+authorTwitter: arielmashraki
+image: https://entgo.io/images/assets/sqlcomment/share.png
+---
+
+One of the common questions we get from the Ent community is how to synchronize objects or references between the
+database backing an Ent application (e.g. MySQL or PostgreSQL) with external services. For example, users would like
+to create or delete a record from in their CRM when a user is created or deleted in Ent, publish a message to a
```
Comment (hilakashai) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:12]:
```suggestion
to create or delete a record from within their CRM when a user is created or deleted in Ent, publish a message to a
```

Comment (hilakashai) [reply doc/website/blog/2021-11-1-sync-objects-in-external-database.md:12]: or 'from their CRM'

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+[Pub/Sub system](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) when an entity is updated, or verify
+references to blobs in object storages such as AWS S3 or Google Cloud Storage.
+
+Ensuring consistency between two separate data systems is not a simple task.  When we want to propagate, for example,
+deletion of a record in one system to another, there is no obvious way to guarantee that the two systems will end in a
+synchronized state, since one of them may fail and the network link between them may be slow or down. Having said that,
+and especially with the prominence of microservices-architectures these problems have become more common, and
```
Comment (hilakashai) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:19]:
```suggestion
and especially with the prominence of microservices-architectures, these problems have become more common, and
```

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+---
+
+One of the common questions we get from the Ent community is how to synchronize objects or references between the
+database backing an Ent application (e.g. MySQL or PostgreSQL) with external services. For example, users would like
+to create or delete a record from within their CRM when a user is created or deleted in Ent, publish a message to a
+[Pub/Sub system](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) when an entity is updated, or verify
+references to blobs in object storages such as AWS S3 or Google Cloud Storage.
+
+Ensuring consistency between two separate data systems is not a simple task.  When we want to propagate, for example,
+deletion of a record in one system to another, there is no obvious way to guarantee that the two systems will end in a
```
Comment (rotemtam) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:17]: The deletion maybe ?

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md (hunk 3 above)
Comment (rotemtam) [reply doc/website/blog/2021-11-1-sync-objects-in-external-database.md:23]: Should a "perfect" design or "perfect" system

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+[Saga Pattern](https://microservices.io/patterns/data/saga.html).
+
+The application of these patterns is usually complex and difficult, and so in many cases architects do not go after a
+"perfect" and instead go after simpler solutions that involve either the acceptance of some inconsistency between the systems or background reconciliation procedures.
+
+In this post, we will not discuss how to solve distributed transactions or implement the Saga pattern with Ent.
+Instead, we will limit our scope to study how to hook into Ent mutations before and after they occur, and run our
+custom logic there.
+
+### Propagate Mutations to External Systems
```
Comment (rotemtam) [inline doc/website/blog/2021-11-1-sync-objects-in-external-database.md:30]: Propagating or propagation of

**Review by rotemtam — CHANGES_REQUESTED** [review]
Not sure about the post title. Let's discuss

File: doc/website/blog/2021-11-1-sync-objects-in-external-database.md (hunk 1 above)
Comment (a8m) [reply doc/website/blog/2021-11-1-sync-objects-in-external-database.md:128]: 👍

File: doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+}
+```
+
+Next, re-run code generation:
+
+```shell
+go generate ./ent
+```
+
+We can now can access the Bucket API from all generated builders. Let's finish the implementations of the above hooks.
```
Comment (hedwigz) [inline doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md:182]:
```suggestion
We can now access the Bucket API from all generated builders. Let's finish the implementations of the above hooks.
```

File: doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+	if err := client.Schema.Create(ctx); err != nil {
+		log.Fatal("failed creating schema resources:", err)
+	}
+	if err := client.User.Create().SetName("a8m").SetAvatarURL("a8m.png").Exec(ctx); err == nil {
+		log.Fatal("expect user creation to fail because the image does not exist in the bucket")
+	}
+	if err := bucket.WriteAll(ctx, "a8m.png", []byte{255, 255, 255}, nil); err != nil {
+		log.Fatalf("failed uploading image to the bucket: %v", err)
+	}
+	fmt.Printf("%q\n", keys(ctx, bucket))
```
Comment (hedwigz) [inline doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md:276]: It's easy enough to understand what `keys` does, but I wonder if you left it out intentionally

**Review by hedwigz — APPROVED** [review]
Great piece :rock:

File: doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md (hunk 15 above)
Comment (a8m) [reply doc/website/blog/2021-11-1-sync-to-external-data-systems-using-hooks.md:276]: Yes

**Comment by a8m** [comment]
Thanks all for the review 🙏 ❤️

**Pull request merged**